[WIP] CASSANDRA-15232 Avoid result truncate in decimal operations

kornilova203 opened the pull request:
See [CASSANDRA-15232](https://issues.apache.org/jira/browse/CASSANDRA-15232)

add (+), subtract (-) and multiply (*) operations:
* before:
  * precision of result used to be always 34 (see MathContext.DECIMAL128)
* after:
  * precision (number of significant digits) of result is at most 10000.
    If result exceeds given precision it will be rounded using HALF_UP mode

division (/) operation:
* before:
  * precision used to be always 34 (see MathContext.DECIMAL128)
* after:
  * expected scale is set to minimum precision (32) minus estimated position of first digit in quotient
  * scale should be at least as big as maximum scale of operands
  * scale should not be less than 32
  * scale should not be bigger than 1000
  * if actual quotient scale is bigger than calculated scale then result is rounded using HALF_UP mode
  * trailing zeros are stripped

modulo (%) operation:
* before:
  * ArithmeticException used to occur when implicit division produces number with precision bigger
    than 34 (see MathContext.DECIMAL128)
* after:
  * No exception

To do:
- [x] do not truncate mod operation
- [x] add tests for mod operation

kornilova203 commented on test/unit/org/apache/cassandra/cql3/functions/OperationFctsTest.java:
```diff
@@ -177,35 +177,35 @@ public void testSingleOperations() throws Throwable
-                       new BigDecimal("0.7857142857142857142857142857142857"),
-                       new BigDecimal("0.9285714285714285714285714285714286"),
+                       new BigDecimal("0.785714285714285714285714285714286"),
+                       new BigDecimal("0.928571428571428571428571428571429"),
```
This change is because precision used to be 34 (in this case it causes 34 significant digits after dot) and now minimum scale is 32 (32 digits after dot)

kornilova-l commented on test/unit/org/apache/cassandra/cql3/functions/OperationFctsTest.java:
```diff
@@ -456,10 +457,16 @@ public void testSingleOperationsWithLiterals() throws Throwable
+                   row(0.15384615384615385, 0.3076923076923077, 0.46153846153846156, 0.6153846153846154, 0.8461538461538461, 1.0, new BigDecimal("1.07692307692307692307692307692308"), new BigDecimal("1.30769230769230769230769230769231")));
+
+        assertRows(execute("SELECT h / 200000000000000000000000000000000000 FROM %s WHERE a = 1 AND b = 2"),
+                   row(new BigDecimal("0.0000000000000000000000000000000000425")));
```
This shows that results are not stripped by minimum scale

kornilova203 commented on test/unit/org/apache/cassandra/cql3/functions/OperationFctsTest.java:
```diff
@@ -456,10 +457,16 @@ public void testSingleOperationsWithLiterals() throws Throwable
+                   row(new BigDecimal("0.0000000000000000000000000000000000425")));
+
+        assertRows(execute("SELECT h / 3 FROM %s WHERE a = 2 AND b = 3"),
+                   row(new BigDecimal("3333.33333333333333333333333333333333")));
```
This test would fault in previous version because precision here is 36 (used to be always 34)

kornilova203 commented on test/unit/org/apache/cassandra/cql3/functions/OperationFctsTest.java:
```diff
@@ -663,7 +670,7 @@ public void testWithDivisionByZero() throws Throwable
                                   "SELECT g / a FROM %s WHERE a = 0 AND b = 2");
 
-        assertInvalidThrowMessage("the operation 'decimal / tinyint' failed: Division by zero",
+        assertInvalidThrowMessage("the operation 'decimal / tinyint' failed: BigInteger divide by zero",
```
Exception arises from different place during division operation

kornilova203 committed:
Avoid result truncate in decimal operations

add (+), subtract (-) and multiply (*) operations:
* before:
  * precision of result used to be always 34 (see MathContext.DECIMAL128)
* after:
  * precision (number of significant digits) of result is at most 10000.
    If result exceeds given precision it will be rounded using HALF_UP mode

division (/) operation:
* before:
  * precision used to be always 34 (see MathContext.DECIMAL128)
* after:
  * expected scale is set to minimum precision (32) minus estimated position of first digit in quotient
  * scale should be at least as big as maximum scale of operands
  * scale should not be less than 32
  * scale should not be bigger than 1000
  * if actual quotient scale is bigger than calculated scale then result is rounded using HALF_UP mode
  * trailing zeros are stripped

modulo (%) operation:
* before:
  * ArithmeticException used to occur when implicit division produces number with precision bigger
    than 34 (see MathContext.DECIMAL128)
* after:
  * No exception

patch by Liudmila Kornilova; reviewed by Benedict Elliott Smith for CASSANDRA-15232

kornilova203 commented:
Merged d60e7988736ed4358595e9c781b110a5bbb5f812
